File: vscode-extension/src/rules/security/sensitiveLoggingRule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SensitiveLoggingRule = void 0;
const finding_1 = require("../../models/finding");
const ruleHelpers_1 = require("../ruleHelpers");
class SensitiveLoggingRule {
    constructor() {
        this.code = 'sensitive-logging';
    }
    evaluate(context) {
        const findings = [];
        for (const file of context.appDartFiles) {
            for (const stmt of (0, ruleHelpers_1.collectLogStatements)(file)) {
                if (!this._looksSensitive(stmt.argument)) {
                    continue;
                }
                findings.push(new finding_1.Finding({
                    severity: finding_1.FindingSeverity.high,
                    confidence: finding_1.FindingConfidence.high,
                    category: finding_1.FindingCategory.security,
                    code: this.code,
                    message: 'Sensitive auth or user data is being logged',
                    fix: 'Remove the log statement or redact auth/session fields before writing anything to logs.',
                    risk: 'Auth tokens and session objects in logs can be harvested from log files, crash reporters, or device storage.',
                    filePath: file.relativePath,
                    line: stmt.startLine,
                }));
            }
        }
        return findings;
    }
    _looksSensitive(argument) {
        const compact = argument.replace(/\s/g, '');
        if (['session', 'token', 'currentUser', 'currentSession'].includes(compact)) {
            return true;
        }
        if (/\$\{?\s*(session|accessToken|refreshToken|currentUser|currentSession|jwt|idToken)\b/i.test(argument)) {
            return true;
        }
        const stripped = argument
            .replace(/"(?:[^"\\]|\\.)*"/g, '')
            .replace(/'(?:[^'\\]|\\.)*'/g, '');
        return /(?<!\w)(session|accessToken|refreshToken|jwt|idToken|currentUser|currentSession|authState|authorization)(?!\w)|\.(currentUser|currentSession|accessToken|refreshToken|idToken)\b/i
            .test(stripped);
    }
}
exports.SensitiveLoggingRule = SensitiveLoggingRule;

File: lib/src/rules/security/sensitive_logging_rule.dart
import '../../models/finding.dart';
import '../../models/project_context.dart';
import '../../rule.dart';
import '../rule_helpers.dart';

class SensitiveLoggingRule extends Rule {
  const SensitiveLoggingRule();

  @override
  String get code => 'sensitive-logging';

  @override
  List<Finding> evaluate(ProjectContext context) {
    final findings = <Finding>[];

    // Use appDartFiles: already excludes test/, integration_test/, example/.
    for (final file in context.appDartFiles) {
      // collectLogStatements handles both single-line and multi-line calls.
      for (final stmt in collectLogStatements(file)) {
        if (!_looksSensitive(stmt.argument)) {
          continue;
        }

        findings.add(
          Finding(
            severity: FindingSeverity.high,
            confidence: FindingConfidence.high,
            category: FindingCategory.security,
            code: code,
            message: 'Sensitive auth or user data is being logged',
            fix:
                'Remove the log statement or redact auth/session fields before writing anything to logs.',
            risk:
                'Auth tokens and session objects in logs can be harvested from log files, crash reporters, or device storage.',
            filePath: file.relativePath,
            line: stmt.startLine,
          ),
        );
      }
    }

    return findings;
  }

  bool _looksSensitive(String argument) {
    // Exact single-identifier matches (compact form strips spaces).
    final compact = argument.replaceAll(' ', '');
    if (compact == 'session' ||
        compact == 'token' ||
        compact == 'currentUser' ||
        compact == 'currentSession') {
      return true;
    }

    // Interpolated sensitive variables: $session, ${accessToken}, etc.
    // The word-boundary prevents matching $username, $userCount, $sessionIndex.
    if (RegExp(
      r'''\$\{?\s*(session|accessToken|refreshToken|currentUser|currentSession|jwt|idToken)\b''',
      caseSensitive: false,
    ).hasMatch(argument)) {
      return true;
    }

    // Interpolated PII variables: $email, ${phoneNumber}, $password, etc.
    if (RegExp(
      r'''\$\{?\s*(email|password|passwd|phoneNumber|phone|ssn|socialSecurity|creditCard|cardNumber|cvv|dateOfBirth|dob)\b''',
      caseSensitive: false,
    ).hasMatch(argument)) {
      return true;
    }

    // Strip string literals so text inside quoted strings doesn't create
    // false positives.  Non-greedy to handle: 'label: ' + accessToken
    final strippedArgument = argument
        .replaceAll(RegExp(r'"(?:[^"\\]|\\.)*"'), '')
        .replaceAll(RegExp(r"""'(?:[^'\\]|\\.)*'"""), '');

    // After stripping, look for identifiers that are exclusively auth-related.
    // Deliberately excludes bare `user` (too broad — matches userCount, etc.)
    if (RegExp(
      r'''(?<!\w)(session|accessToken|refreshToken|jwt|idToken|currentUser|currentSession|authState|authorization)(?!\w)'''
      r'''|\.(?:currentUser|currentSession|accessToken|refreshToken|idToken)\b''',
      caseSensitive: false,
    ).hasMatch(strippedArgument)) {
      return true;
    }

    // PII identifiers: email, password, phone, SSN, credit card, etc.
    if (RegExp(
      r'''(?<!\w)(password|passwd|creditCard|cardNumber|cvv|ssn|socialSecurity|dateOfBirth)(?!\w)'''
      r'''|\.(?:password|creditCard|cardNumber|ssn|socialSecurity)\b''',
      caseSensitive: false,
    ).hasMatch(strippedArgument)) {
      return true;
    }

    return false;
  }
}
